### ISSUE TITLE
Better error handling - Is it possible to add "if exception" operator

### ISSUE THREAD

**Issue by hydro63** [dart-lang/language#4030]
I was wondering why there isn't a "if exception" operator. This operator would have similar functionality as "??" operator, but instead of taking right side on null, it takes right side on exception. It would replace this:
```dart
String out;
try {
    out = funcThatThows();
} catch (e) {
    out = "";
}
```
with this:
```dart
var out = funcThatThrows() !! "";
```
It could also be chained with "??" operator
```dart
String? nullOrThrow(){
    if(...) throw Exception("Problem!");
    if(...) return null;
    return "Hello";
}

var out = nullThrow() !! "Exception occured" ?? "Returned null";

print(out) // prints "Hello" or "Exception occured" or "Returned null";
```
I've also seen syntax where `try` is an expression
```dart
var out = try funcThatThrows() else "Hello";
```
Here is a more real world example of where it could be used:
```dart
class VTTResult{
    ...
}

class VTTParser{
    ...
    static VTTResult parse(String input){
        ....
        else {
            throw FormatException("Invalid Token found");
        }
    }
}

void loadFile(File inputFile){
    var loadedStructure = VTTParser.parse(inputFile.readAsStringSync()) !! VTTResult();
}
```
Will it be ever added or is there a reason against adding it?

**Comment by jakemac53** [comment]
Not what you are asking for, and the ergonomics aren't as great, but fwiw you can use extensions to create wrappers of functions which will return some value on error, and gets you kind of close, like https://dartpad.dev/?id=972ccebd83d92cef57c656557211d959

**Comment by hydro63** [comment]
After using it for some time, i've found that its not really a solution and it only adds its own troubles - its difficult to read the code where its used, and for every function that has different kind of paramaters i need to duplicate it for it to work properly

As such, i propose adding **one** of these features:
1) add if-exception operator as used in `var out = funcThatThrows() !! "";`
2) add "try ... else ..." expression as used in `var out = try funcThatThrows() else "Hello";`
3) allow to be used for retreating the thrown exception - `var x = catch funcThrows(); // thrown -> Exception, else -> value`

By adding one of them, it would be possible to quicky and efficiently convert thrown Exceptions into some default object / configuration. Combined with the addition of union types (which ASAIK are in the process of specification), it would be possible for the developper to choose to use exception as value. It would also be possible to convert said. 
```dart
int funcThrows() ...

// variable type -> int | Exception
var exceptionOrValue = funcThrows() !! Exception();
var exceptionOrValue2 = try funcThrows() else Exception();
// the best in my opinion
var exceptionOrValue3 = catch funcThrows(); // if throws returns the thrown error, else return value
```
Combined with pattern matching, it could be used as:
```dart
if(funcThrows() case var value when value is Exception){
...
}
```

**Comment by jakemac53** [comment]
Yeah I do agree it is a bit weird and doesn't read as nicely as a built in feature.

You could do it with a regular function instead, and always wrap the expression in a closure, but performance might become a possible concern, https://dartpad.dev/?id=1962a5f03068730878a5967c7f126edd. This pattern can is I think more intuitive ultimately though.

One advantage of the other pattern is that you get a new function back, which you can pass around to other places, which is kind of cool. But, the specialization per function type is annoying to be sure.

**Comment by hydro63** [comment]
With the rising popularity of errors as values (i know its not entirely correct), I feel like it would be a great addition to the language, since it would enable the developper to choose when to use it as value, and when to do a more complex error handling. The biggest reason why many people choose to not handle errors, is mainly because of the amount of work it requires. By adding this syntactic sugar, those people would handle most errors.
In order to make error handling more used, a better syntax is needed.
```dart
// simple exception handing - from experience 90% of the time
var value = try funcThatThrows() else "Exception found";

// more complex exception handing
try {
    var value = funcThatThrows();
    ...
} on Exception catch(e) {
    // complex handling of the exception
    ...
}
```
I am advocating the `try ... else ...` syntax, because forbids complex chaining of the handling (difficult to read), which should be handled in `try ... catch ...` block. Also, this syntax would only work on Exceptions, not on Errors (we don't want to catch those).
In terms of errors as values, here is comparison with go
```go
f, err := funcThatThrows()
if err != nil {
    log.Fatal(err)
}
```
```dart
var value = try funcThatThrows() else Exception();
if(value is Exception){
    print("Problem occured");
}
```
While the use of errors as values is not the main part of this proposal (that being easier syntax to handle it), its a sideeffect that has good implications

**Comment by hydro63** [comment]
I like the syntax you propose, though I wouldn't allow omitting the "try" keyword, because it would allow more undefined behavior, where you don't know what part actually throws. Meaning
```dart
// catch only b or a + b including the + operator?
a + b catch ...

// with try keyword
// catches a + b
try a + b catch ...

// catches only b
a + try b catch ...
```

**Comment by k757657** [comment]
maps_toolkit: ^3.0.0

**Comment by hydro63** [comment]
I like the syntax you propose, though I wouldn't allow omitting the "try" keyword, because it would allow more undefined behavior, where you don't know what part actually throws. Meaning
```dart
// catch only b or a + b including the + operator?
a + b catch ...

// with try keyword
// catches a + b
try a + b catch ...

// catches only b
a + try b catch ...
```

**Comment by tatumizer** [comment]
More concise syntax may create an illusion that try-catch can be used casually, as if it were equivalent to error handling in other languages (like rust and golang). However, the current implementation of try-catch is *very* slow, which might be quite surprising to the users.

**Comment by hydro63** [comment]
I actually didn't really notice the try-catch speed issue, mainly because i don't code apps where speed is important, but that is indeed a problem. That being said, more concise syntax would also improve the use of error handling, and not just propagating the error till the app breaks.

I'd say that if a function is gonna throw, its better for it to be handled when it's easy to handle it, rather than not, where the exception is gonna crash the program. In my opinion, in most apps its better to prioritize stability - by making the error handling easier and thus making it more used - , rather than speed, which is rarely an issue.

**Comment by HenriqueNas** [comment]
so @hydro63, the `try-catch expressions` would work pretty much like `switch expressions`?

```dart
  String out = try funcThatThows() {
    Argument || Format => "default",
    Exception(message: var m) => m, 
    _ => rethrow, // throws out of this try-catch
  };


  final result = try (() async {
      throw 'error';
    })() {
      _ => 'nothing',
    }
    .then(print) // nothing
    .catchError(print); // never reached

```

**Comment by hydro63** [comment]
that's what i would have wanted, since it would allow for quick error handling where possible

**Comment by HenriqueNas** [comment]
> ˆ
> I feel it would help the error handling, even if by small amount.

I agree a lot, seems like this might increase the average code quality of Dart projects.. at least for those I work on

**Comment by jakemac53** [comment]
Fwiw, you could also use something like https://github.com/eernstg/extension_type_unions to model this, or create a `Result` extension type, like https://dartpad.dev/?id=c9c0a0d6e3f1954b5a125d55a6d2ab7f. 

That gets you pretty close to the desired syntax without language changes:

```dart
var result = switch (Result.of(funcThatThrows)) {
  String s => s,
  Object? e => e.toString(), // Whatever you want to do on error
};
```

And the `Result.of` could be omitted if you wrote all your functions to already return a result type instead of throwing, this example is just showing wrapping of an existing function.

**Comment by hydro63** [comment]
Thanks, that i'll try it out, but so far it seems much more usable than to make extension on specific type.

_Also, i just want to say that this was my first feature request, so it's probably quite shitty for what it's worth_

**Comment by lrhn** [comment]
Could also just allow `catch pattern` in any switch, where `case pattern` could occur, catching errors thrown in the switch expression. Or in expression switches `catch pattern =>`

That would require a non-catch clause to be exhaustive though. 

```dart
var v = switch (expr) {
  var x => x,
  catch FormatException () => 0,
};
```

Could do that and use `try` instead of `switch` for speeches with no non-catch clauses.

Or maybe it's too weird.

**Comment by hydro63** [comment]
I have thought about it a bit more, and come to the conclusion that quick error handling is impossible to do correctly. Here are the problems i found
### Problems
There are essentially two kinds of syntaxes proposed
1) Binary `if - exception` operator - think `!!` or `try func else default`
1) Try expressions

The problem with binary operator syntax is that it presumably catches every exception and you can't really define a type which it catches.

The problem with both try expressions and binary operator, is that there is no indication for the user. 

Let's say we are loading a configuration, and the configuration does not exists, so we get an exception. But since it's just a configuration, you can just provide some default values, and the program will work. The problem is that the user should be notified that they made a mistake, whether it be popup or normal notificafion, he has to know, so that he can solve the issue, if he is the one that caused it. The expression syntax just doesn't allow that. That's a problem, since most easily solvable exceptions are a user's fault (wrong file loaded, wrong url, wrong credentials, insufficient privilages)

Another problem with both of them is that debugging will become really difficult. Ideally, the catched errors should still be printed to the console, just to see if it's programmer's fault or the user's fault.

If we were to print all the errors caught this way, it should not be in a release build. It should be possible to add the compiler option for printing these errors, but that just overcomplicates the issue.

In conclusion, as somebody pointed out, i think making quicker error checking this way is just very problems prone.

### Different direction of error handling
Since going in the direction of making the error handling into expression is very error prone, i thought going in different direction to making error handling more accesible is needed. 

But before that, let's understand what we want to achieve. Here are my requirements:
1) Error handling should be easier to write, and understand
1) Make it so that the developper uses error handling more often.

The features it would be good to have:
1) New way of error handling should be expressive yet concise

Drawing inspiration from [JS error handling proposal](https://github.com/arthurfiorette/proposal-safe-assignment-operator), i propose adding similar safe assignment operator `?=`, that would transform the return value of a function into either union `T | Exception`, record `(Exception? err, T? value), or into `Result<T>` type.

The benefits would be exactly the same as in the JS version, with it being more or less optional `errors as values`. The key thing being **optional**, meaning the users will use it if they like this kind of error handling.

Now, i know it's already possible to do this with `Result<T>` type, but i think that the record `(Exception err, T value)` is a better way to wrap the errors than `Result<T>` type. That is because i can quickly reassing a default value to `T value`, without making another variable to hold it

Example:
```dart
// with record (Exception? err, Config? conf)
var (Exception? err, Config? conf) ?= getConfiguration(file); //can throw

if(err != null){
  conf = defaultConf;
  notify(err); //will notify the user of the error
}

//with Result<Config>
var _config ?= getConfiguration(file);
late Config conf;
if(_config.isError){
  conf = defaultConf;
  notify(_config.error);
} else {
  conf = _config.value;
}
```
_PS. I haven't yet worked with Result type in Dart, but i assume that's how it works?_

**Comment by lrhn** [comment]
The result class is indeed just two subclasses, one for an error and over for a value. Will be sealed when migrated to Dart 3.

My problems with a type like `(Object? error, T? value)` are: 
- it's not disjoint. The type allows both to be non-`null`. And both to be `null`.
- Cannot distinguish "no value" from the `null` value of `T` is nullable.
- You need to promote the value to non-`null`.

So it requires an interpretation: 
- if the error is non-null, ignore the value, which will be `null`.
- if the error is `null`, assume the `value` is a `T`. You still need to do an `as T` cast on it to get rid of the `?`.

A switch checking `(var error?, _)` and `(null, var value as T)` should be exhaustive. I don't know if we can figure that out, but it _can throw at runtime_ if both values are `null`. 
That's also the minimum checks that you need to write.
For `Result`, at least if it becomes sealed, the switch cases are `Error e` and `Value(:var value)`. And exhaustiveness should make that only be one type check.

No matter which type is used for capturing error-or-value, I like the idea of an error capturing operator. I don't think it needs to be linked to assignment, it can just work on any expression.

Maybe make it `try(expr)`, which has static type `ValueOrError<T, Object>` (strawman name) if `expt` has static type `T`, and which catches anything thrown by `expr`.
Maybe allow `try<Exception>(expr)` to only catch `Exception`s, and let errors stay thrown. 

Or make it a suffix operator, like a selector: `expr.catch` or `expr.catch<Exception>`.

Then we can always discuss what `typedef ValueOrError<T, E extends Object> = ...` should be shaped like.
(I don't see a good way to make the catch operator user definable, because it's not really a member of the type. 
(Maybe we can make it a special extension-only operator. Hmm. Let's not.)

**Comment by hydro63** [comment]
I don't want to try doing `try` expressions, (at least the ones with `switch` like behaviour), because i can't think of any pattern that would fulfill all the basic requirements of new error handling i've outlined. It will either become more difficult to write than actual try statements or will have serious sideeffects.

> My problems with a type like (Object? error, T? value) are:

I'm also not a real fan of the `Record` return, but i found it better than `Result<T>`, because it would allow me to write more concise code, without declaring dummy variables. The best solution would be `Exception | T`, where it would automatically promote the type to `T` after the user already took care of the `Exception` possibility or assigned default value to it.

My current requirements of the new error handling pattern are this:
1) Error handling should be easier to write, and understand
1) Make it so that the developper uses error handling more often.
1) The typing should be sound, and null safe.

Also if you have want to change the requirements of the new error handling, or you have some addition, we can change them. But at least for now, if the new pattern doesn't fulfill these requirements, i don't think we should accept it.

**Comment by lrhn** [comment]
A plain untagged union type isn't really useful when the possible thrown values are `Object` and the possible result values are `Object?`. It really needs to be a disjoint union, so at least tagged, and then it might as well be a real sealed class hierarchy.

What you can do with a `Result` object depends on the API we give it... and as a type, which extensions people put on it.

If the `Result<T>` class has a `T? get value` and an `Object? get error`, then it is _as_ useful as the record type.
You can do `result case Result(:var value, :var error)` just as for a record. (And if we can make the `Result` implicit, it would be almost as short.)

But then the `Value<T>` subclass can have `T get value` and `Null get error`, so you can check whether it's a value first, and then not need to cast the accessed value. By not being a record type, just an interface type, it can ensure that a result is either a value or an error, never both or neither. 

So still leaning towards a `Result` sealed class API as the most flexible and still viable approach. (Even if I don't particularly like introducing magical classes that the runtime and language depends on. This will be something similar to a `Future`, capturing the result of a computation, just without the asynchrony.)

**Comment by hydro63** [comment]
If `Result<T>` type got some nice API, i would have absolutely no problem with it being used as the return value. But it at least has to have at least this:
```dart
//  return value if not error, else return default value
extension ResultAddition<T> on Result<T>{
  T resolve(T defaultValue){
    ...
  }
}
T resolve(T defaultValue);

// you can use it as such
func().catch().resolve(15);
```
This functionality is not difficult to implement for the user, but i feel it needs to be part of the actual type.

Apart from that it would be nice (not required) to have some mechanism for quickly changing `Result<T>` into value without losing expressiveness we can achieve with `try ... catch`. `try catch` would still be used when the error is too difficult to quickly solve, but the user could also do something like this _(from previous example)_, just with `Result<T>` instead of `Record`:
```dart
// try to achieve this with Result<T> instead of Record
var (Exception? err, Config? conf) = getConfiguration(file).catch(); //can throw

if(err != null){
  conf = defaultConf;
  notify(err); //will notify the user of the error
}
```
One way it could be possible, is if the `switch` expression were allowed to have multiple statements (#3117), it would be solved really easily
```dart
var conf = switch(getConfiguration(file).catch()){
  Result(:var value) => value,
  Result(:var error) {
    notify(error);
    defaultValue
  }
};
```
This is not really essential, and is more of a _what if_, but it would be useful to have.
_PS. What i want to achieve is to make the Result type more usable, not redefining the whole error handling_
